Refactor to add kernel boot param config for Pi (#49)
canonical/real-time-ubuntu-docs: docs/how-to/modify-kernel-boot-parameters.rst

How to modify kernel boot parameters

> =======================================

Reviewer: Nitpick: remove the extra "===".

Author: Lovely ReST.

The real-time performance-tuning features compiled into the kernel can be
controlled by various means. The approach used here is a common one:
passing command line parameters via a bootloader.

The Linux kernel supports a wide range of parameters, but only a few of them are relevant to real-time tuning.
:doc:`/reference/kernel-boot-parameters` is good a reference.

Depending on the underlying boot mechanism, there are multiple ways to pass command line parameters to the kernel.

> For Raspberry Pi, jump to :ref:`raspberry-pi`.

Reviewer: ```suggestion
- For Raspberry Pi, jump to :ref:`raspberry-pi`.
- Ubuntu Core uses GRUB but provides a different interface for...
```

Suggesting to indent this for readability because you mentioned "multiple ways to pass command line parameters" and these are the 2 methods you are referring to.

Author: I agree with the poor readability, but I don't feel bullets make it better:

![image](https://github.com/canonical/real-time-ubuntu-docs/assets/11150423/94732468-a6fa-438d-b28c-94011934b046)

vs.

![image](https://github.com/canonical/real-time-ubuntu-docs/assets/11150423/ad2e2095-bb8e-49a6-a700-5d3574ac8ba3)

What do you think?

Reviewer: I think the second option is better. The bullets (or rather the indentation) makes it a bit more obvious that these are the 3 different methods, depending on your system.

thanks!

Author: Ok, changed. Also added a link to the grub section.

Ubuntu Core uses GRUB but provides a different interface for 
`modifying kernel options <https://ubuntu.com/core/docs/modify-kernel-options>`_.

.. Replace the Ubuntu Core link to the internal how-to (https://documentation.ubuntu.com/real-time/en/ubuntu-core-devel/how-to/uc-boot-parameters/), once released.

Continue reading for systems that use the GNU GRUB bootloader.

Modify kernel boot parameters via GRUB
--------------------------------------

You can temporarily or permanently configure parameters which GRUB will pass to
Real-time Ubuntu.

When first testing a real-time tuning feature, `configure it temporarily
<#configure-grub-temporarily>`_. Temporary bootloader parameters are passed to
the kernel on a single startup. To re-use them you'll need to re-enter them ---
they don't persist across reboots.

When you're satisfied that a parameter achieves what you want, you can make it
permanent by `updating GRUB's configuration <#configure-grub-permanently>`_
then rebooting.

Configure GRUB temporarily
~~~~~~~~~~~~~~~~~~~~~~~~~~

Reboot, and interrupt GRUB's startup sequence at its bootloader menu.

> Highlight the entry --- usually **Ubuntu** --- that would boot Real-time

Reviewer: Hey @farshidtz , is there a reason why you use 3 `---` instead of just a single dash `-`?
Thanks.

Author: I'm just refactoring this text, moving it from the CPU config doc to the modifying kernel config. The `---` creates an Em dash:
![image](https://github.com/canonical/real-time-ubuntu-docs/assets/11150423/d2ea7906-2ff3-4ec8-8d31-193d4fa92615)

Reviewer: Oh, as long as the em dash was intentional.
thanks!

Ubuntu.

.. image:: grub-menu.png
   :alt: GRUB menu
   :width: 100%
   :align: center

> Press **e** to edit its boot parameters.

Author: The kbd markup was used a few lines below, but not here. Not sure why. As said, I'm just moving text :D

> Append parameters to the line that starts with ``linux``:

Reviewer: ```suggestion
Append parameters to the lines that start with ``linux``:
```

.. image:: grub-edit.png
   :alt: GRUB edit
   :width: 100%
   :align: center

* Leave a space before each new parameter

* Don't add space round ``=`` and other punctuations for each key-value

> * Don't add line breaks

Reviewer 2: I don't think it's necessary to repeat these instructions three times. It would be better to add a `note` section before introducing the temporary GRUB configuration, stating that it is valid for the following configurations.

Reviewer: I agree it's a bit repetitive and should be refactored somehow.

Also modifying kernel boot parameters isn't something just anyone would be doing; we should be able to expect a certain level of competency from the user doing this. And in that case the example (with at least 2 parameters) should be sufficient.

But I am OK to leave this out of this PR for now.

Reviewer: On a related note, I tried comparing what other docs did for this.
Surprisingly none of them actually indicated the editing rules for these files.

For reference:
https://wiki.ubuntu.com/Kernel/KernelBootParameters#Permanently_Add_a_Kernel_Boot_Parameter
https://wiki.archlinux.org/title/kernel_parameters
https://wiki.archlinux.org/title/kernel_parameters#GRUB

Author: I think it is good to mention these directly when the user is making the changes. A note somewhere on top could be forgotten. Adding a section and referring to it is also not ideal because it breaks the flow for something as small as this. 

With that said, the text is mostly copied from cpu config doc to another one. I prefer making separate PRs to make improvements that are outside the scope of this PR: adding Pi config.

Author: @AnneCYH I wrote the above comment before seeing your replies.

Let's consider these in another PR.

> Surprisingly none of them actually indicated the editing rules for these files.

The Ubuntu wiki has the following, which is even more detailed:

> Press SPACE to add a blank space (after "splash") then carefully type in the kernel boot parameter that you need to add. (If you need to add multiple parameters separate them with SPACE but do not add spaces before or after any = signs or punctuation in the parameters themselves).

Press :kbd:`Ctrl+X` to exit the editor and continue booting with the new
parameters.

Configure GRUB permanently
~~~~~~~~~~~~~~~~~~~~~~~~~~

> Once your device has booted, launch use a text editor to edit ``/etc/default/grub``.

Reviewer: ```suggestion
Once your device has booted, use a text editor to edit ``/etc/default/grub``.
```

> Append parameters to the line that starts with ``GRUB_CMDLINE_LINUX``:

Reviewer: Hey @farshidtz , sorry, i know you said you're refactoring so this is kinda 'out of scope'.

But wanted to check with you about this line.

If i open the file on my machine there are 2 lines that **start** with `GRUB_CMDLINE_LINUX`.
> GRUB_CMDLINE_LINUX_DEFAULT="quiet splash"
> GRUB_CMDLINE_LINUX="log_buf_len=16M drm.debug=0x1e"

From what I understand (and Googled!) I think it should only be appended to the last line. If that's the case this should probably be worded to be more specific.

E.g. "Append parameters to the line that starts with ``GRUB_CMDLINE_LINUX=``.

Author: Both should work, the instructions originally referred to `GRUB_CMDLINE_LINUX_DEFAULT` but I changed it in this PR to not edit the defaults. I guess the whole working is strange because the parameters aren't appended, but rather inserted inside the quotations. 

Changed.
